### DISCUSSION TITLE
How should uncertainty work in v4

### DISCUSSION THREAD

**Discussion by paulf81** [NatLabRockies/floris#777] [v4 Design Discussion]
Hi @misi9170, @rafmudaf, @bayc  and @Bartdoekemeijer (and anyone else interested to weigh in!), I was hoping to start a discussion now on how we think uncertainty should work in v4.  This connects to Issue #725 , that uncertainty doesn't work in v3 with time series =True, and in v4, that "time-series" structure is pervasive.  I believe though we can use the WindRose tool currently being build out in pull request #775 to maintain the same pattern, if we want.

But I was thinking it could be nice to think about restructuring a set of related tools spanning FLORIS and FLASC, built up to take advantage of v4's structure.  I have in mind:

- FLORIS
  - UncercaintyInterface
  - ParallelInterface
- FLASC
  - floris_tools
    - calc_floris_approx_table
    - add_gaussian_blending_to_floris_approx_table

I was imagining all of these capabilities could be built from the same building blocks, something like:

-  Determine cases needed to be run in order to:
  - Add gaussian blur
  - Establish an approx table
- Use hashing or similar to recognize cases that are already run
- Parallelization isn't hard tied to this problem, but seems worth tackling at once?

Thought we could use this discussion to propose what this will look like so we have something in mind when we reach this step

**Comment by misi9170** [comment]
I think your suggestions above make sense @paulf81. Just to make sure I'm understanding correctly, I'm adding a little bit of detail to what I understand each step in our process above to be:

1. Characterize the set of findex conditions that have been passed to `calculate_wake` to determine an "envelope" of conditions needed to add blurring/uncertainty. For example, if the user has specified N wind conditions that are all at 8 m/s and with wind directions varying between 200 and 250 degrees, we would set the conditions to _actually_ run as, say, wind directions in one degree steps between 190 and 260 degrees, all with the wind speed of 8 m/s. With these, a n_findex_orig sets of Gaussian weights will be computed according to how each of the originally-specified findex conditions matches the new set of findex conditions. These Gaussian weights will depend on `wd_std`.

2. `reinitialize()` with the set of wind needed conditions (likely using a `WindData` object) and run `calculate_wake()`.

3. Use the previously-calculated Gaussian weights to map the computer powers (or any other variable of interest) to the conditions provided by the user.

This way, we don't really need the hashing---that is basically handled by the construction of the set of conditions that will be actually run, and mapped back to what the user asked for by the n_findex_orig sets of Gaussian weights.

**Reply by paulf81** [reply]
I think that is basically right.  Maybe we also just need a little bit of extra logic in the case where the user passes in a TimeSeries, and wd/ws are not yet binned, where we bin/unbin the data, or maybe immediatly do a to_windrose operation and return the result based on that?  That could work nicely actually I think

**Comment by misi9170** [comment]
Implemented in #821
